File: problems/0019-remove-nth-node-from-end-of-list/python/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def build(vals):
    dummy = ListNode()
    cur = dummy
    for v in vals:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


def to_list(head):
    out = []
    while head:
        out.append(head.val)
        head = head.next
    return out


def removeNthFromEnd(head, n):
    """Remove Nth Node From End of List — https://leetcode.com/problems/remove-nth-node-from-end-of-list/"""
    dummy = ListNode(0, head)
    fast = slow = dummy
    for _ in range(n):
        fast = fast.next
    while fast.next:
        fast = fast.next
        slow = slow.next
    slow.next = slow.next.next
    return dummy.next


def removeNthFromEndArr(vals, n):
    return to_list(removeNthFromEnd(build(vals), n))

File: problems/0019-remove-nth-node-from-end-of-list/typescript/solution.ts
class ListNode {
  val: number;
  next: ListNode | null = null;
  constructor(val = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function build(vals: number[]): ListNode | null {
  const dummy = new ListNode();
  let cur = dummy;
  for (const v of vals) {
    cur.next = new ListNode(v);
    cur = cur.next;
  }
  return dummy.next;
}

function toArray(head: ListNode | null): number[] {
  const out: number[] = [];
  for (let n = head; n !== null; n = n.next) out.push(n.val);
  return out;
}

// Remove Nth Node From End of List — https://leetcode.com/problems/remove-nth-node-from-end-of-list/
function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
  const dummy = new ListNode(0, head);
  let fast: ListNode = dummy;
  let slow: ListNode = dummy;
  for (let i = 0; i < n; i++) {
    fast = fast.next as ListNode;
  }
  while (fast.next !== null) {
    fast = fast.next;
    slow = slow.next as ListNode;
  }
  slow.next = slow.next !== null ? slow.next.next : null;
  return dummy.next;
}

export function removeNthFromEndArr(vals: number[], n: number): number[] {
  return toArray(removeNthFromEnd(build(vals), n));
}
